drivers: serial: silabs: Implement PM and Async API with DMA to silabs eusart driver

Martinhoff-maker opened the pull request:
The goal of this PR is to add three functionality to the silabs uart-eusart driver:
- Asynchronous API with the use of DMA.
- (PM) Introduce RESUME and SUSPEND actions. 
- (PM) Introduce deep sleep prevention with get/put pm states lock

yishai1999 commented on drivers/serial/Kconfig.silabs_eusart:
```diff
@@ -8,8 +9,22 @@ config UART_SILABS_EUSART
+
+if UART_SILABS_EUSART
+
+config UART_SILABS_EUSART_ASYNC
```
Why do you need this?  Isn't `CONFIG_UART_ASYNC_API` enough?

Martinhoff-maker replied:
no because this symbol is shared between uart driver but we need to know if our driver can support async. see (https://github.com/zephyrproject-rtos/zephyr/pull/85914).

jerome-pouiller commented on drivers/serial/uart_silabs_eusart.c:
```diff
@@ -205,16 +237,504 @@ static void eusart_irq_callback_set(const struct device *dev, uart_irq_callback_
 
-static void eusart_isr(const struct device *dev)
+#ifdef CONFIG_UART_SILABS_EUSART_ASYNC
+static inline void eusart_async_user_callback(struct eusart_data *data,
```
Don't use `inline` in `.c` files (`static inline` is used in the header files to avoid gcc complains about unused function).

This comment applies to all the functions below.

Martinhoff-maker replied:
Actually, the conversion function from ll to cfg (eusart_ll2cfg_stopbits in example) are not used so removing the "inline" will cause a warning. I wonder to delete them but it might be useful in the future.

jerome-pouiller commented on drivers/serial/uart_silabs_eusart.c:
```diff
@@ -205,16 +237,504 @@ static void eusart_irq_callback_set(const struct device *dev, uart_irq_callback_
+}
+
+static int eusart_async_callback_set(const struct device *dev, uart_callback_t callback,
+						 void *user_data)
```
Indentation

jerome-pouiller commented on drivers/serial/uart_silabs_eusart.c:
```diff
@@ -205,16 +237,504 @@ static void eusart_irq_callback_set(const struct device *dev, uart_irq_callback_
+}
+
+void eusart_dma_rx_cb(const struct device *dma_dev, void *user_data, uint32_t channel,
+				  int status)
```
Indentation.

jerome-pouiller commented on drivers/serial/uart_silabs_eusart.c:
```diff
@@ -205,16 +237,504 @@ static void eusart_irq_callback_set(const struct device *dev, uart_irq_callback_
+{
+	const struct device *uart_dev = user_data;
+	struct eusart_data *data = uart_dev->data;
+	struct uart_event disabled_event = {.type = UART_RX_DISABLED};
```
```
struct uart_event disabled_event = {
	.type = UART_RX_DISABLED
};
```

jerome-pouiller commented on drivers/serial/uart_silabs_eusart.c:
```diff
@@ -205,16 +237,504 @@ static void eusart_irq_callback_set(const struct device *dev, uart_irq_callback_
+		return;
+	}
+
+	(void)k_work_cancel_delayable(&data->dma_rx.timeout_work);
```
I don't believe the `(void)` is required.

jerome-pouiller commented on drivers/serial/uart_silabs_eusart.c:
```diff
@@ -205,16 +237,504 @@ static void eusart_irq_callback_set(const struct device *dev, uart_irq_callback_
+}
+
+void eusart_dma_tx_cb(const struct device *dma_dev, void *user_data, uint32_t channel,
+				  int status)
```
... and so on

jerome-pouiller commented on drivers/serial/Kconfig.silabs_eusart:
```diff
@@ -8,8 +9,23 @@ config UART_SILABS_EUSART
+config UART_SILABS_EUSART_ASYNC
+	bool
+	depends on DMA_SILABS_LDMA
+	depends on UART_ASYNC_API
```
It does not make sense to enable DMA if `UART_ASYNC_API` is not used?

asmellby replied:
For code size reasons, since UART only uses DMA in the async API, we should IMO avoid compiling with DMA support if async API is disabled. UART is different from SPI and I2C, where DMA can also be used for blocking transfers, since UART as 3 separate APIs for single-character polling, interrupt-driven transfers and async transfers.

jerome-pouiller reviewed (commented):
The last commit is a bit tricky. For now, I have only focused o the first ones.

jerome-pouiller reviewed (commented):
Most of my comments are cosmetics. The code is very good.

Martinhoff-maker committed:
drivers: serial: silabs: introduce silabs eusart runtime configure

Introduce runtime configuration API for the silabs eusart driver.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
tests: driver: uart: add xg23_rb4210a board overlay

Add xg23_rb4210a board overlay to test asynchronous api with eusart
driver

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: serial: silabs: introduce pm for silabs eusart

Make use of pm_device_driver_init to perform driver initialization.
Implement PM suspend and resume, which performs the following actions:

* Enables/disables the USART
* Gates the USART clock
* Configures USART pins

Also take PM locks to prevent deep sleep during TX and RX operations.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: serial: silabs: replace prefix in silabs eusart driver

Replace prefix from "uart_silabs_eusart" to "eusart" in order to
simplify the code, and make it more readable.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: serial: silabs: harmonize silabs eusart code with silabs usart

Apply code cleaning done in silabs usart driver to silabs eusart driver.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: serial: silabs: correction of silabs eusart code indentation

Apply correct indentation to silabs eusart driver.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: serial: silabs: split init in silabs eusart driver

Split of uart_silabs_eusart_init function to have better visibility of
initialization.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: serial: silabs: introduce asynchronous silabs eusart

Introduce silabs asynchronous eusart with dma support.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

jhedberg commented:
@Martinhoff-maker did you forget to push an update? I can see you marked most of Jérôme's comments from yesterday as resolved, however there hasn't been any update to this PR for a couple of weeks.

Martinhoff-maker commented:
@jhedberg I just push my code today because I see some problems after rebasing on top on the main branch.

jerome-pouiller commented:
@yishai1999, we need your approval to be able to merge this PR. Is it okay for you?
